### PR TITLE
Improve compat infrastructure

### PR DESCRIPTION
**Pull request by proux01**
Implements what has been decided during Coq Call: https://github.com/coq/coq/wiki/Coq-Call-2024-08-27#handling-of--compat-infrastructure

#### Overlay (to be merged in sync with the PR)

* https://github.com/ejgallego/coq-serapi/pull/426

### CODE REVIEW

**Comment by Zimmi48** [comment]
The main argument in favor of this was that this compat infrastructure was no longer used.

In a [Coq Call discussion](https://github.com/coq/coq/wiki/Coq-Call-2024-08-27), it was decided to keep this infrastructure, but make it more useful by:

- only warning and not failing when a compat version is used which is no longer supported (similar to the behavior of flags and options)
- using these compat files to silence deprecation warnings

**Comment by proux01** [comment]
Sorry, I'm still working on it, please let me a few weeks. I updated the title.

**Comment by proux01** [comment]
CI green, this seems ready

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (Zimmi48) [inline doc/sphinx/practical-tools/coq-commands.rst:422]: How is it going to work with the renaming to Rocq? Ideally, compat files should still be names `Coq819`, `Coq820` and `Rocq90`, no?

File: doc/changelog/09-cli-tools/19370-compat-warning.rst
Comment (Zimmi48) [inline doc/changelog/09-cli-tools/19370-compat-warning.rst:2]:
```suggestion
  The ``-compat`` :ref:`command line option <command-line-options>` raises
```

File: doc/changelog/09-cli-tools/19370-compat-warning.rst
Comment (Zimmi48) [inline doc/changelog/09-cli-tools/19370-compat-warning.rst:4]: Maybe describe the reason for this change in case the users do not think of it by themselves.

BTW, what about a third changelog entry to note (or noting here) that compat files now silence deprecation warnings that were introduced since then?

File: sysinit/coqinit.ml
Comment (Zimmi48) [inline sysinit/coqinit.ml:189]: "Cannot require" is not very clear. "Did not find" or something like that would be clearer IMHO. And since this warning is only raised when using `-compat` options, what about "Did not find compatibility module..."?

File: theories/Compat/README.md
Comment (Zimmi48) [inline theories/Compat/README.md:4]: That's already the most important info to include in this README, but I would have hoped for a little bit more explanations about this infrastructure here.

**Review by Zimmi48 — APPROVED** [review]
Looks great to me!

Minor comments below (the most important one being that we need to anticipate the move to `Rocq`). The previous infrastructure had the advantage that, by behind ad-hoc, it would have made this simpler. Maybe it can be decided whether to use the `Coq` or the `Rocq` prefix based on whether it starts by `8.` or a bigger number?

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (proux01) [reply doc/sphinx/practical-tools/coq-commands.rst:422]: My idea was simply to rename all files to Rocq* even if Rocq 8.18 has never been really a thing. Otherwise we can use you idea. Anyway since the command line option is `-compat X.Y` the precise name of the file doesn't play much of a role.

Comment (proux01) [reply doc/sphinx/practical-tools/coq-commands.rst:422]: I'd prefer to do the Coq -> Rocq renaming in a further small PR that I'll open just after this one, in order to avoid mixing unrelated changes (thr current PR has nothing to do with the renaming itself).

Comment (Zimmi48) [reply doc/sphinx/practical-tools/coq-commands.rst:422]: > Anyway since the command line option is `-compat X.Y` the precise name of the file doesn't play much of a role.

Ah, that's a good point (less overengineered than my solution). We may still want to keep `Coq81X` modules just in case some people where directly depending on them though.

But then why not do this renaming of compat files to `RocqXY` in this PR as well?

Comment (Zimmi48) [reply doc/sphinx/practical-tools/coq-commands.rst:422]: > I'd prefer to do the Coq -> Rocq renaming in a further small PR that I'll open just after this one, in order to avoid mixing unrelated changes (thr current PR has nothing to do with the renaming itself).

OK, but keep in mind that the second PR will have to update the changelog of this PR.

**Review by Zimmi48 — APPROVED** [review]
Do we want tests that the option behaves as it should when the file does not exist? And when it does?

Do we need other reviewers to look at this before merging?

**Comment by proux01** [comment]
> Do we want tests that the option behaves as it should when the file does not exist? And when it does?

I added a test (for each case)

**Comment by Zimmi48** [comment]
> Do we need other reviewers to look at this before merging?

I will merge at the end of the week if there are no more comments.

File: test-suite/success/compat_not_found.v
Comment (Zimmi48) [inline test-suite/success/compat_not_found.v:1]: If you put this file in `output` instead of `success`, you will be able to check that the warning was emitted.

Comment (proux01) [reply test-suite/success/compat_not_found.v:1]: Why not, done

**Comment by Zimmi48** [comment]
@coqbot merge now

**Pull request merged**
